### PR TITLE
Allow the Controller and Webhook Containers to run with the securityContext: readOnlyRootfilesystem: true

### PR DESCRIPTION
**Pull request by npgretz**
### 🛑 Important:
Please open an issue to discuss significant work before you start. We appreciate your contributions and don't want your efforts to go to waste!

For guidelines on how to contribute, please review the [CONTRIBUTING.md](CONTRIBUTING.md) document.

## Purpose of this PR
Provide a clear and concise description of the changes. Explain the motivation behind these changes and link to relevant issues or discussions.

Close [#2218 ](https://github.com/kubeflow/spark-operator/issues/2218#issue-2565007558)

**Proposed changes:**
- Mount an emptyDir onto the controller container to write Spark artifacts to
- Mount an emptyDir onto the webhook container to write certficates to
- Set the default securityContexts for the controller and webhook to readOnlyRootfilesystem: true

## Change Category
Indicate the type of change by marking the applicable boxes:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that could affect existing functionality)
- [ ] Documentation update

### Rationale

Please see this [issue](https://github.com/kubeflow/spark-operator/issues/2218#issue-2565007558) for discussion around this feature.
This feature improves the default security posture of the Spark Operator Helm Chart and allows organizations that require `readOnlyRootfilesystem: true` to deploy the helm chart.

## Checklist
Before submitting your PR, please review the following:

- [X] I have conducted a self-review of my own code.
- [X] I have updated documentation accordingly.
- [X] I have added tests that prove my changes are effective or that my feature works.
- [X] Existing unit tests pass locally with my changes.

### CODE REVIEW

**Comment by ChenYi015** [comment]
@npgretz Could you do a rebase as there is a merge conflict?

**Comment by npgretz** [comment]
> @npgretz Could you do a rebase as there is a merge conflict?

Done. Thank you for making me aware!

**Comment by npgretz** [comment]
Sorry for the small mistakes holding up the merge; this is my first open-source contribution. I am just figuring out the process.

File: charts/spark-operator-chart/values.yaml [hunk 1]
```diff
@@ -105,7 +105,11 @@ controller:
     # key2: value2
 
   # -- Volumes for controller pods.
-  volumes: []
+  volumes:
+    # -- Create a tmp directory to write Spark artifacts to for deployed Spark apps.
```
Comment (ChenYi015) [inline charts/spark-operator-chart/values.yaml:109]:
```suggestion
    # Create a tmp directory to write Spark artifacts to for deployed Spark apps.
```

Remove the `--` so this comment will not rendered in the `README.md` file.

File: charts/spark-operator-chart/values.yaml [hunk 2]
```diff
@@ -141,7 +145,11 @@ controller:
   envFrom: []
 
   # -- Volume mounts for controller containers.
-  volumeMounts: []
+  volumeMounts:
+    # -- Mount a tmp directory to write Spark artifacts to for deployed Spark apps.
```
Comment (ChenYi015) [inline charts/spark-operator-chart/values.yaml:149]:
```suggestion
    # Mount a tmp directory to write Spark artifacts to for deployed Spark apps.
```

The same as above.

File: charts/spark-operator-chart/values.yaml [hunk 3]
```diff
@@ -252,7 +261,11 @@ webhook:
   sidecars: []
 
   # -- Volumes for webhook pods.
-  volumes: []
+  volumes:
+    # -- Create a dir for the webhook to generate its certificates in.
```
Comment (ChenYi015) [inline charts/spark-operator-chart/values.yaml:265]:
```suggestion
    # Create a dir for the webhook to generate its certificates in.
```

The same.

File: charts/spark-operator-chart/values.yaml [hunk 4]
```diff
@@ -288,7 +301,13 @@ webhook:
   envFrom: []
 
   # -- Volume mounts for webhook containers.
-  volumeMounts: []
+  volumeMounts:
+    # -- Mount a dir for the webhook to generate its certificates in.
```
Comment (ChenYi015) [inline charts/spark-operator-chart/values.yaml:305]:
```suggestion
    # Mount a dir for the webhook to generate its certificates in.
```

**Review by ChenYi015 — COMMENTED** [review]
It would be better to use the K8s yaml style for `values.yaml` file, i.e.， the array need not be indented.

**Comment by ChenYi015** [comment]
@npgretz Thanks for the contribution to enhance the security of the Spark operator. Just a few small nits, could you address them and regenerate the helm README by running `make helm-docs`? Then everything will be fine for me.

**Comment by npgretz** [comment]
> It would be better to use the K8s yaml style for `values.yaml` file, i.e.， the array need not be indented.

Thank you for your suggestions! Could ask for clarification on this suggestion. Is this referring to the array of objects for the volumes and volumeMounts I added for the controller and webhook pods? I am not sure how I could create these arrays of objects without the indentation and have valid YAML.

I also did not find a reference for K8's YAML Style. Could you direct me to a resource?

Thank you, again!

File: charts/spark-operator-chart/values.yaml (hunk 1 above)
Comment (npgretz) [reply charts/spark-operator-chart/values.yaml]: I completed these changes locally and pushed them but now realize I could've directly committed these changes from your suggestions.

File: charts/spark-operator-chart/values.yaml (hunk 2 above)
Comment (npgretz) [reply charts/spark-operator-chart/values.yaml]: Done

File: charts/spark-operator-chart/values.yaml (hunk 4 above)
Comment (npgretz) [reply charts/spark-operator-chart/values.yaml]: Done

File: charts/spark-operator-chart/values.yaml (hunk 3 above)
Comment (npgretz) [reply charts/spark-operator-chart/values.yaml]: Done

**Comment by ChenYi015** [comment]
@npgretz If you run `kubectl get pod -o yaml` to fetch the manifest of any pod, you will find the array objects are not indented:

```
$ kubectl get pod -n spark-operator spark-operator-controller-69844fbfb5-nqwts -o yaml
...
spec:
  containers:
  - args:
    - controller
    - start
    - --zap-log-level=info
    - --namespaces=default
    - --controller-threads=10
    - --enable-ui-service=true
 ...
```

> Is this referring to the array of objects for the volumes and volumeMounts I added for the controller and webhook pods?

Yes, you can simply remove the indentation so that we can maintain a consistent YAML style.

**Comment by npgretz** [comment]
@ChenYi015, thank for you for explaining that. I hadn't realized the format of the templated manifests was off. I have adjusted the deploy templates for the controller and webhook to correctly indent volumes and volumeMounts.

**Review by ChenYi015 — APPROVED** [review]

**Pull request merged**